jinja2/filters.py

```diff
@@ -518,9 +518,17 @@ def do_int(value, default=0, base=10):
     can also override the default base (10) in the second
     parameter, which handles input with prefixes such as
     0b, 0o and 0x for bases 2, 8 and 16 respectively.
+    The base is ignored for non-string values.
     """
+    # only string types support int() with explicit base
+    if isinstance(value, string_types) and base != 10:
+        try:
+            return int(value, base)
+        except ValueError:
+            return default
+
     try:
-        return int(value, base)
+        return int(value)
     except (TypeError, ValueError):
         # this quirk is necessary so that "42.23"|int gives 42.
```

Comment:
With your fix, this quirk may get skipped when required. I prefer a solution where logic is not duplicated.